pkg/background/generate/generate.go

```diff
@@ -413,13 +432,341 @@ func getResourceInfoForDataAndClone(rule kyvernov1.Rule) (kind, name, namespace,
 	return
 }
 
+func forEachGetResourceInfoForDataAndClone(fe kyvernov1.ForEachGeneration) (kind, name, namespace, apiversion string, err error) {
+	if len(fe.CloneList.Kinds) == 0 {
+		if kind = fe.Kind; kind == "" {
+			return "", "", "", "", fmt.Errorf("%s", "kind can not be empty")
+		}
+		if name = fe.Name; name == "" {
+			return "", "", "", "", fmt.Errorf("%s", "name can not be empty")
+		}
+	}
+	namespace = fe.Namespace
+	apiversion = fe.APIVersion
+	return
+}
+
+func forEach(log logr.Logger, client dclient.Interface, rule kyvernov1.Rule, resource unstructured.Unstructured, ctx enginecontext.EvalInterface, policy kyvernov1.PolicyInterface, ur kyvernov1beta1.UpdateRequest, fe kyvernov1.ForEachGeneration, element interface{}) ([]kyvernov1.ResourceSpec, error) {
+	var noGenResource kyvernov1.ResourceSpec
+	var newGenResources []kyvernov1.ResourceSpec
+	rdatas := []GenerateResponse{}
+	var cresp, dresp map[string]interface{}
+	var mode ResourceMode
+
+	tempFe, err := substituteAllInForEach(fe, ctx, log)
+	fe = *tempFe
+	if err != nil {
+		return newGenResources, nil
+	}
+
+	genKind, genName, genNamespace, genAPIVersion, err := forEachGetResourceInfoForDataAndClone(fe)
```

Comment:
Why was this done in a separate function when the function is only used once and all data is available locally?